ITISFoundation/osparc-simcore
# ✨🗃️ Implements services listing for nih-portal (1/2) (⚠️  devops)

## What do these changes do?

Implements ``/service`` entrypoint to list osparc services metadata. The responses also provides a ``view_url`` link to create a study with the latest version of the selected service. The same link can be used to append a file to the service.
Note that links produce by ``/v0/services`` will not be operative until next PR.

### Changes Overview

- [x] ✨ new table that keeps latest versions of all services
- [x] catalog background task updates table 
- [x] ~~catalog ``list_services`` API call new query filters ``only_latest: bool``, ``include_classifiers``~~
  - Too complex via catalog service. Created shortcut in ``studies_dispatcher._catalog`` that directly reads the db
- [x] ``/services`` calls ``list_services(only_latest=True, include_classifiers={ "osparc" } )``
    - [x] transform response to ``ServiceGet``
    - [x] create a ``view_url`` for each of them

To simplify the review and have a running version in master ASAP (so it can be tested against the portal dev by @ignapas ) the remaining features will be included in a separate  PR

### Following PR 
- Add mimetypes (currently empty) 
- Modify implementation of  ``/view`` redirection (given in ``view_url``) 
        -  open study with service
        -  open study service with file


### ⚠️ devops

- ⚠️ new env var w/ default ``STUDIES_DEFAULT_SERVICE_THUMBNAIL=https://via.placeholder.com/170x120.png``
- 🗃️ new table ``services_latest``: automatically updated by the catalog's service with a periodic background task


## Related issue/s

- related to ITISFoundation/osparc-issues#689
- related to ITISFoundation/osparc-issues#683


## How to test

- Deploy stack
- Check in the database that new table ``services_latest`` was filled after a few moments after startup
- ``GET /v0/services`` should respond as
<img width="1424" alt="Screen Shot 2023-03-27 at 01 18 40" src="https://user-images.githubusercontent.com/32402063/227811031-c9a045e0-f78c-4e8c-a4a3-c45b79b50bd8.png">
